grudge/symbolic/operators.py

```diff
@@ -651,20 +658,41 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+        from meshmode.discretization import ElementGroupWithBasis
+
+        n = volgrp.order
+        m = afgrp.order
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+        face_quad_degree = 2*max(n, m)
+
+        for iface, face in enumerate(faces):
+            try:
+                face_quadrature = afgrp._quadrature_rule()
```

Comment:
It's not amazing that we have to reach into an underscored property here.

Reply:
So the simplest solution I can see here is to just rename `_quadrature_rule()` to `quadrature_rule()`. That seems fine to me.

On a related note: any comment on whether you think this is appropriate? The rationale for this try-except clause is that if a `afgrp` has a quadrature rule already associated with it, then we should just use that rather than create a new quadrature object. Thoughts?

Reply:
The `unit_nodes` and `weights` are public methods on the element groups, so you could also just yank those and stick them into a `mp.Quadrature` directly.

Not sure if that's less hacky or more hacky though..

Reply:
I see what you mean. But it feels wasteful to keep instantiating new `mp.Quadrature` objects when there is one already available.

My second paragraph is more or less just stating that I'm thinking we should try to use the quadrature rule defining the `afgrp` (if there is one) rather than just keep creating new quadrature objects.

Reply:
That's a fair point. I don't feel too bad about it because the results from this get memoized, so it's not a big cost.

Reply:
OK as long as you all don't think it's too much of an issue, I can be convinced it's okay :smile: